File: backend/api/routes/verification.py
from fastapi import APIRouter, Depends, HTTPException, status

from backend.api.dependencies import get_verification_service
from backend.api.schemas.verification import (
    VerificationRequest,
    VerificationResponse,
)
from backend.services.verification import VerificationService

router = APIRouter(
    prefix="/sessions/{session_id}/verification",
    tags=["verification"],
)


@router.post("", response_model=dict)
async def verify_patient(
    session_id: str,
    request: VerificationRequest,
    service: VerificationService = Depends(get_verification_service),
) -> dict:
    try:
        verification_id, result, existing_patient = (
            await service.verify(
                session_id,
                request.method,
                request.identifier,
            )
        )
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

    visit_type = (
        "RETURNING_VISIT"
        if existing_patient is not None
        else "FIRST_VISIT"
    )

    response = VerificationResponse(
        verification_id=verification_id,
        status=result.status,
        patient_id=(
            existing_patient.patient_id
            if existing_patient is not None
            else None
        ),
        existing_patient=existing_patient is not None,
        visit_type=visit_type,
    )

    return {"data": response.model_dump(mode="json")}


@router.get("", response_model=dict)
async def get_verification(
    session_id: str,
    service: VerificationService = Depends(get_verification_service),
) -> dict:
    try:
        verification_id, verification_status, patient_id = (
            await service.get_status(session_id)
        )
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

    existing_patient = patient_id is not None

    response = VerificationResponse(
        verification_id=verification_id,
        status=verification_status,
        patient_id=patient_id,
        existing_patient=existing_patient,
        visit_type=(
            "RETURNING_VISIT"
            if existing_patient
            else "FIRST_VISIT"
        ),
    )

    return {"data": response.model_dump(mode="json")}

File: patient-web/src/api/verification.ts
import { apiRequest } from "./client";

export interface VerificationOtpChallenge {
  challenge_id: string;
  masked_destination: string;
  expires_in_seconds: number;
  demo_otp: string | null;
}

export interface PatientIdentificationResponse {
  verification_id: string;
  status: string;
  patient_id: string | null;
  existing_patient: boolean;
  visit_type: "FIRST_VISIT" | "RETURNING_VISIT";
}

interface IdentificationRequest {
  method: string;
  identifier: string;
}

interface VerificationOtpRequest {
  challenge_id: string;
  otp: string;
}

export async function requestIdentityOtp(
  sessionId: string,
  method: string,
  identifier: string,
): Promise<VerificationOtpChallenge> {
  const response = await apiRequest<{
    data: VerificationOtpChallenge;
  }>(`/sessions/${encodeURIComponent(sessionId)}/verification/otp`, {
    method: "POST",
    body: JSON.stringify({
      method,
      identifier,
    } satisfies IdentificationRequest),
  });

  return response.data;
}

export async function verifyPatientOtp(
  sessionId: string,
  challengeId: string,
  otp: string,
): Promise<PatientIdentificationResponse> {
  const response = await apiRequest<{
    data: PatientIdentificationResponse;
  }>(`/sessions/${encodeURIComponent(sessionId)}/verification`, {
    method: "POST",
    body: JSON.stringify({
      challenge_id: challengeId,
      otp,
    } satisfies VerificationOtpRequest),
  });

  return response.data;
}
